driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -119,17 +119,28 @@ private void startNewExecution() {
     }
 
     private void scheduleExecution(long delayMillis) {
-        if (isDone.get() || delayMillis <= 0)
+        if (isDone.get() || delayMillis < 0)
             return;
         if (logger.isTraceEnabled())
             logger.trace("[{}] Schedule next speculative execution in {} ms", id, delayMillis);
-        scheduledExecutions.add(scheduler.newTimeout(newExecutionTask, delayMillis, TimeUnit.MILLISECONDS));
+        if(delayMillis == 0) {
+            // kick off request immediately
+            try {
+                newExecutionTask.run(null);
+            } catch (Exception e) {
+                // the newExecutionTask shouldnt be able to throw a non-runtime but pokecatch needed for netty api
```

Comment:
Was wondering if we could improve this. Maybe if you extract some code out of the timeout task body:

```
private final TimerTask newExecutionTask = new TimerTask() {
    @Override
    public void run(final Timeout timeout) {
        scheduledExecutions.remove(timeout);
        scheduleExecutionImmediately();
    }
};

private void scheduleExecution(long delayMillis) {
    if (isDone.get() || delayMillis < 0)
        return;
    if (logger.isTraceEnabled())
        logger.trace("[{}] Schedule next speculative execution in {} ms", id, delayMillis);
    if (delayMillis == 0) {
        // kick off request immediately
        scheduleExecutionImmediately();
    } else {
        scheduledExecutions.add(scheduler.newTimeout(newExecutionTask, delayMillis, TimeUnit.MILLISECONDS));
    }
}

private void scheduleExecutionImmediately() {
    if (!isDone.get())
        // We could be the timer thread, so reschedule to another executor
        manager.executor().execute(new Runnable() {
            @Override
            public void run() {
                if (metricsEnabled())
                    metrics().getErrorMetrics().getSpeculativeExecutions().inc();
                startNewExecution();
            }
        });
}

```
Besides I wonder if we could avoid using `manager.executor()` to execute the speculative attempts when the delay is zero, given that we are not on an internal thread. Although sending the speculative query on the same thread as the initial one would have a funny side-effect: the speculative request is likely to be sent before the first (non-speculative) request. Not sure if this is of a concern.